pkg/config/config.go

```diff
@@ -195,13 +204,68 @@ func evalFuncFile() function.Function {
 	})
 }
 
-// LoadComponentConfigBody returns nil if no component with the given
-// name is found in the configuration
-func (c *Config) LoadComponentConfigBody(componentName string) *hcl.Body {
-	for _, component := range c.RootConfig.Components {
-		if componentName == component.Name {
-			return &component.Config
+// lokocfgFilenames returns a slice containing the names of all .lokocfg files found under path p.
+// If p is not a directory, p is returned by itself in the slice.
+func lokocfgFilenames(p string) ([]string, error) {
+	isDir, err := util.PathIsDir(p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat config path %q: %w", p, err)
+	}
+
+	if !isDir {
+		return []string{p}, nil
+	}
+
+	globPattern := filepath.Join(p, "*.lokocfg")
+
+	names, err := filepath.Glob(globPattern)
+	if err != nil {
+		return nil, fmt.Errorf("bad filepath glob pattern %q: %w", globPattern, err)
+	}
+
+	return names, nil
+}
+
+// populateVars accepts a slice of variables and a map of user-provided values. The function
+// populates each variable with the corresponding user-provided value or with a default. If a
+// variable doesn't have a matching value and there is no default, an error is returned (as
+// hcl.Diagnostics).
+func populateVars(vars []variable, values map[string]cty.Value) (map[string]cty.Value, hcl.Diagnostics) {
+	res := map[string]cty.Value{}
+
+	for _, v := range vars {
+		if value, ok := values[v.Name]; ok {
+			res[v.Name] = value
+
+			continue
+		}
+
+		if len(v.Default) == 0 {
+			continue
+		}
+
+		defaultValue, hasDefaultValue := v.Default["default"]
+		if !hasDefaultValue {
+			continue
+		}
+
+		dv, diags := defaultValue.Expr.Value(nil)
+		if diags.HasErrors() {
+			return nil, diags
 		}
+
+		res[v.Name] = dv
+	}
+
+	return res, nil
+}
+
+// diag returns an hcl.Diagnostics with a single hcl.Diagnostic based on the provided error string.
+func diag(err string) hcl.Diagnostics {
```

Comment:
Maybe it would be better to accept `error` here, rather than `string`, as we always return diagnostic with `Severity: hcl.DiagError` ?
```suggestion
func diag(err error) hcl.Diagnostics {
```

Reply:
Yes, somehow I thought we'd have a problem in cases where we want to format the error string, but we can just wrap with `fmt.Errorf()` before calling `diag()`. Fixed.